lib/xdrgen/generators/go.rb

```diff
@@ -549,6 +531,259 @@ def render_encode_to_body(out, var, type, self_encode:)
+        out.puts "  return n, nil"
+        out.puts "}"
+        out.break
+      end
+
+      def render_variable_declaration(out, indent, var, type, declared_variables:)
+        if !declared_variables.include?var
+           out.puts "#{indent}var #{var} #{type}"
+           declared_variables.append(var)
+        end
+      end
+
+      # render_decode_from_body assumes there is an `d` variable containing an
+      # xdr.Decoder, and a variable defined by `var` that is the value to
+      # encode.
+      def render_decode_from_body(out, var, type, declared_variables:, self_encode:)
+        tail = <<-EOS.strip_heredoc
+          n += nTmp
+          if err != nil {
+            return n, err
+          }
+        EOS
+        optional = type.sub_type == :optional
+        if optional
+          render_variable_declaration(out, "  ", 'b', "bool", declared_variables: declared_variables)
+          out.puts "  b, nTmp, err = d.DecodeBool()"
+          out.puts tail
+          out.puts "  #{var} = nil"
+          out.puts "  if b {"
+          out.puts "     #{var} = new(#{name type})"
+        end
+        case type
+        when AST::Typespecs::UnsignedHyper
+          out.puts "  #{var}, nTmp, err = d.DecodeUhyper()"
+          out.puts tail
+        when AST::Typespecs::Hyper
+          out.puts "  #{var}, nTmp, err = d.DecodeHyper()"
+          out.puts tail
+        when AST::Typespecs::UnsignedInt
+          out.puts "  #{var}, nTmp, err = d.DecodeUint()"
+          out.puts tail
+        when AST::Typespecs::Int
+          out.puts "  #{var}, nTmp, err = d.DecodeInt()"
+          out.puts tail
+        when AST::Typespecs::String
+          arg = "0"
+          arg = type.decl.resolved_size if !type.decl.resolved_size.nil?
+          out.puts "  #{var}, nTmp, err = d.DecodeString(#{arg})"
+          out.puts tail
+        when AST::Typespecs::Opaque
+          if type.fixed?
+            out.puts "  nTmp, err = d.DecodeFixedOpaqueInplace(#{var}[:])"
+            out.puts tail
+          else
+            arg = "0"
+            arg = type.decl.resolved_size if !type.decl.resolved_size.nil?
+            out.puts "  #{var}, nTmp, err = d.DecodeOpaque(#{arg})"
+            out.puts tail
+          end
+        when AST::Typespecs::Simple
+          case type.sub_type
+          when :simple, :optional
+            optional_within = type.is_a?(AST::Identifier) && type.resolved_type.sub_type == :optional
+            if optional_within
+              render_variable_declaration(out, "  ", 'b', "bool", declared_variables: declared_variables)
+              out.puts "  b, nTmp, err = d.DecodeBool()"
+              out.puts tail
+              out.puts "  #{var} = nil"
+              out.puts "  if b {"
+              out.puts "     #{var} = new(#{name type.resolved_type.declaration.type})"
+            end
+            var = "(*#{name type})(#{var})" if self_encode
+            out.puts "  nTmp, err = #{var}.DecodeFrom(d)"
```

Comment:
❔  I noticed in stellar/go#4064 that this generates the following code where `SponsoringId` is a `type SponsorshipDescriptor *AccountId`:
```go
		s.SponsoringId = new(AccountId)
		nTmp, err = s.SponsoringId.DecodeFrom(d)
```
I'm surprised you can call `DecodeFrom` on a `SponsorshipDescriptor` since that type has no functions. I would have thought you'd need to either dereference `s.SponsoringId` first so that you have an `AccountId` or convert it to `*AccountId`.

I see stellar/go#4064 passed its tests though, so I assume it compiles and works. I don't understand why though because `SponsorshipDescriptor` shouldn't inherit any functions.

If you know why, please share.

Reply:
Uhm, good point. I think it just uses the `*AccountId` method.

Reply:
How though. `SponsorshipDescriptor` is its own type, it doesn't inherit methods. The type of the value can be converted back to *AccountId, or it can magically become it through dereferencing to `AccountId`, but I don't see how this works. 🤔 

For example, the following code fails to compile.

```go
package main

type AccountId int

func (a AccountId) Method() {}

type SponsorshipDescriptor *AccountId

func main() {
	var s SponsorshipDescriptor
	s = new(AccountId)
	s.Method()
}
```
Ref: https://play.golang.org/p/kTxwVGPMe4X

As far as I can tell the type of `s.SponsoringId` is `SponsorshipDescriptor`.

Reply:
Your example is not accurate, it's missing an equal, Sponsoring descriptor is an alias.

See https://play.golang.org/p/LBAb4oxe2vm

Reply:
See 

```
        if typedef.sub_type == :optional
          out.puts "type #{name typedef} = #{reference typedef.declaration.type}"
        else
          out.puts "type #{name typedef} #{reference typedef.declaration.type}"
        end
```

Reply:
Of course! Haha, [I was the one who changed it to be an alias](https://github.com/stellar/xdrgen/commit/92a28e1f3ef3a2b7e9409a2baa2c5ed0751d3e1c#diff-559f5809716dacbc678a75e354ead997243db1419c9186a359d5ebf749fed6e2R28-R31) and I forgot 🤦🏻.